source/fuzz/fuzzer_pass_wrap_vector_synonym.cpp

```diff
@@ -0,0 +1,158 @@
+
+        // Populate components based on vector type and size.
+        for (uint32_t i = 0; i < vector_size; ++i) {
+          if (i == position) {
+            vec1_components.emplace_back(target_id1);
+            vec2_components.emplace_back(target_id2);
+          } else {
+            if (type->AsInteger()) {
+              // Operands are integers. Fill other positions with zero
+              // integer constants.
+              vec1_components.emplace_back(FuzzerPass::FindOrCreateZeroConstant(
+                  FuzzerPass::FindOrCreateIntegerType(width,
+                                                      is_signed_constant),
+                  true));
+              vec2_components.emplace_back(FuzzerPass::FindOrCreateZeroConstant(
+                  FuzzerPass::FindOrCreateIntegerType(width,
+                                                      is_signed_constant),
+                  true));
+            } else if (type->AsFloat()) {
+              // Operands are floats. Fill other positions with zero
+              // float constants.
+              vec1_components.emplace_back(FuzzerPass::FindOrCreateZeroConstant(
+                  FuzzerPass::FindOrCreateFloatType(width), true));
+              vec2_components.emplace_back(FuzzerPass::FindOrCreateZeroConstant(
+                  FuzzerPass::FindOrCreateFloatType(width), true));
+            } else {
+              assert(
+                  false &&
+                  "Instruction opcode must be a valid numeric constant type.");
+            }
+          }
+        }
+        // Add two OpCompositeConstruct to the module with result id returned.
+        // Add the first OpCompositeConstruct that wraps the id of the first
+        // operand.
+        uint32_t result_id1 = GetFuzzerContext()->GetFreshId();
+        ApplyTransformation(TransformationCompositeConstruct(
+            operand_type_id, vec1_components, instruction_descriptor,
+            result_id1));
+
+        // Add the second OpCompositeConstruct that wraps the id of the second
+        // operand.
+        uint32_t result_id2 = GetFuzzerContext()->GetFreshId();
+        ApplyTransformation(TransformationCompositeConstruct(
+            operand_type_id, vec2_components, instruction_descriptor,
+            result_id2));
+
+        // Add synonym facts between original operands and id from the
+        // |scalar_position| of the vectors are added.
+        GetTransformationContext()->GetFactManager()->AddFactDataSynonym(
+            MakeDataDescriptor(result_id1, {position}),
+            MakeDataDescriptor(instruction_iterator->GetSingleWordInOperand(0),
+                               {}));
+
+        GetTransformationContext()->GetFactManager()->AddFactDataSynonym(
+            MakeDataDescriptor(result_id2, {position}),
+            MakeDataDescriptor(instruction_iterator->GetSingleWordInOperand(1),
+                               {}));
```

Comment:
I believe this is redundant and possibly dangerous.

I think it's redundant because `TransformationCompositeConstruct` should already have added these synonyms.

And I think it's dangerous because it might be the case that either/both of target_id_1 or target_id_2 have the "id is irrelevant" fact associated with them. In this case it's illegal for them to participate in a synonym.

I suggest you replace these lines with assertions, along the lines of:

```
assert(( is_irrelevant(target_id_1) || is_synonymous(target_id_1, DataDescriptor(result_id1, {position})) && "Synonym fact should have been added already.")
```

(where you'll need to query the fact manager for `is_irrelevant` and `is_synonymous`)

and similarly for the other operand.

I can explain a bit more about this during our next sync if it would help.

Reply:
Done.